backend/danswer/datastores/vespa/store.py

```diff
@@ -0,0 +1,402 @@
+        f"{DOCUMENT_ID}, "
+        f"{CHUNK_ID}, "
+        f"{BLURB}, "
+        f"{CONTENT}, "
+        f"{SOURCE_TYPE}, "
+        f"{SOURCE_LINKS}, "
+        f"{SEMANTIC_IDENTIFIER}, "
+        f"{SECTION_CONTINUATION}, "
+        f"{BOOST}, "
+        f"{METADATA} "
+        f"from {DOCUMENT_INDEX_NAME} where "
+    )
+
+    def __init__(self, deployment_zip: str = VESPA_DEPLOYMENT_ZIP) -> None:
+        # Vespa index name isn't configurable via code alone because of the config .sd file that needs
+        # to be updated + zipped + deployed, not supporting the option for simplicity
+        self.deployment_zip = deployment_zip
+
+    def ensure_indices_exist(self) -> None:
+        """Verifying indices is more involved as there is no good way to
+        verify the deployed app against the zip locally. But deploying the latest app.zip will ensure that
+        the index is up-to-date with the expected schema and this does not erase the existing index.
+        If the changes cannot be applied without conflict with existing data, it will fail with a non 200
+        """
+        deploy_url = f"{VESPA_APPLICATION_ENDPOINT}/tenant/default/prepareandactivate"
+        headers = {"Content-Type": "application/zip"}
+        with open(self.deployment_zip, "rb") as f:
+            response = requests.post(deploy_url, headers=headers, data=f)
+            if response.status_code != 200:
+                raise RuntimeError(
+                    f"Failed to prepare Vespa Danswer Index. Response: {response.text}"
+                )
+
+    def index(
+        self,
+        chunks: list[IndexChunk],
+        index_attempt_metadata: IndexAttemptMetadata,
+    ) -> set[DocumentInsertionRecord]:
+        return _index_vespa_chunks(
+            chunks=chunks, index_attempt_metadata=index_attempt_metadata
+        )
+
+    def update(self, update_requests: list[UpdateRequest]) -> None:
+        logger.info(
+            f"Updating {len(update_requests)} documents' allowed_users in Vespa"
+        )
+
+        json_header = {"Content-Type": "application/json"}
+
+        for update_request in update_requests:
+            if update_request.boost is None and update_request.allowed_users is None:
+                logger.error("Update request received but nothing to update")
+
+            update_dict: dict[str, dict[str, list[str] | int]] = {"fields": {}}
+            if update_request.boost:
+                update_dict["fields"][BOOST] = update_request.boost
+            if update_request.allowed_users:
+                update_dict["fields"][ALLOWED_USERS] = update_request.allowed_users
+
+            for document_id in update_request.document_ids:
+                for doc_chunk_id in _get_vespa_chunk_ids_by_document_id(document_id):
+                    url = f"{DOCUMENT_ID_ENDPOINT}/{doc_chunk_id}"
+                    res = requests.put(url, headers=json_header, json=update_dict)
+
+                    if res.status_code != 200:
```

Comment:
also fail loudly here if update fails to avoid marking the delete as successful without actually updating all permissions